### ISSUE TITLE
Docker run from a image with same image ID and size behaves differently

### ISSUE THREAD

**Issue by yuchengwu** [moby/moby#39247]
**Description**
Docker run from a image with same image ID and size behaves differently.

**Steps to reproduce the issue:**
On our environment, given the `broken` image not removed from host 516, it can be reliably reproduced,
1. check image `transwarp/node-exporter:live` exists on both 516 and 519, with same id and size
![Selection_114](https://user-images.githubusercontent.com/13599124/58097672-82f32200-7c0a-11e9-9e81-d29430b372da.png)
2. docker run on 516
```
[root@tos-516 ~]# docker run -it --rm --user $(whoami) -v /etc/passwd:/etc/passwd:ro transwarp/node-exporter:live
docker: Error response from daemon: linux spec user: unable to find user root: no matching entries in passwd file. 

[root@tos-516 ~]# docker run -it --rm --user $(id -u) -v /etc/passwd:/etc/passwd:ro transwarp/node-exporter:live
standard_init_linux.go:190: exec user process caused "exec format error"                

[root@tos-516 ~]# docker run --rm -ti transwarp/node-exporter:live  --user root
docker: Error response from daemon: linux spec user: unable to find user nobody: no matching entries in passwd file.    


[root@tos-516 ~]# docker run -it --rm -v /etc/passwd:/etc/passwd:ro transwarp/node-exporter:live                    
docker: Error response from daemon: linux spec user: unable to find user nobody: no matching entries in passwd file.   
```

3. docker run on 519
```
[root@tos-519 ~]# docker run -it --rm --user $(id -u) -v /etc/passwd:/etc/passwd:ro transwarp/node-exporter:live
INFO[0000] Starting node_exporter (version=0.17.0, branch=HEAD, revision=f6f6194a436b9a63d0439abc585c76b19a206b21)  source="node_exporter.go:82"                                                           
INFO[0000] Build context (go=go1.11.2, user=root@322511e06ced, date=20181130-15:51:33)  source="node_exporter.go:83"                                                                                       
INFO[0000] Enabled collectors:                           source="node_exporter.go:90"
INFO[0000]  - arp                                        source="node_exporter.go:97"
INFO[0000]  - bcache                                     source="node_exporter.go:97"
INFO[0000]  - bonding                                    source="node_exporter.go:97"
INFO[0000]  - conntrack                                  source="node_exporter.go:97"
INFO[0000]  - cpu                                        source="node_exporter.go:97"
INFO[0000]  - diskstats                                  source="node_exporter.go:97"
INFO[0000]  - edac                                       source="node_exporter.go:97"
INFO[0000]  - entropy                                    source="node_exporter.go:97"
INFO[0000]  - filefd                                     source="node_exporter.go:97"
INFO[0000]  - filesystem                                 source="node_exporter.go:97"
INFO[0000]  - hwmon                                      source="node_exporter.go:97"
INFO[0000]  - infiniband                                 source="node_exporter.go:97"               
INFO[0000]  - ipvs                                       source="node_exporter.go:97"                
INFO[0000]  - loadavg                                    source="node_exporter.go:97"               
INFO[0000]  - mdadm                                      source="node_exporter.go:97"               
INFO[0000]  - meminfo                                    source="node_exporter.go:97"
INFO[0000]  - netclass                                   source="node_exporter.go:97"
INFO[0000]  - netdev                                     source="node_exporter.go:97"
INFO[0000]  - netstat                                    source="node_exporter.go:97"
INFO[0000]  - nfs                                        source="node_exporter.go:97"
INFO[0000]  - nfsd                                       source="node_exporter.go:97"
INFO[0000]  - sockstat                                   source="node_exporter.go:97"
INFO[0000]  - stat                                       source="node_exporter.go:97"
INFO[0000]  - textfile                                   source="node_exporter.go:97"
INFO[0000]  - time                                       source="node_exporter.go:97"
INFO[0000]  - timex                                      source="node_exporter.go:97"
INFO[0000]  - uname                                      source="node_exporter.go:97"
INFO[0000]  - vmstat                                     source="node_exporter.go:97"
INFO[0000]  - xfs                                        source="node_exporter.go:97"
INFO[0000]  - zfs                                        source="node_exporter.go:97"
INFO[0000] Listening on :9100                            source="node_exporter.go:111"
```

**Describe the results you received:**
images with same `IMAGE ID`  have different behavior

**Describe the results you expected:**
images with same `IMAGE ID` should have same behavior

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue happens occasionally, it has happened twice so far. The first time we observed the two images have same `IMAGE ID` but different image `SIZE`(docker images xxx), also one could run happily but another failed to run.

**Output of `docker version`:**

on 516 
```
[root@tos-516 ~]# docker version 
Client:
 Version:       17.12.1-ce
 API version:   1.35
 Go version:    go1.9.4
 Git commit:
 Built: Fri Mar 29 11:13:12 2019
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.1-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   cdade7ddb4
  Built:        Fri Mar 29 11:16:05 2019
  OS/Arch:      linux/amd64
  Experimental: true
```

on 519
```
[root@tos-519 ~]# docker version 
Client:
 Version:       17.12.1-ce
 API version:   1.35
 Go version:    go1.9.4
 Git commit:
 Built: Fri Mar 29 11:13:12 2019
 OS/Arch:       linux/amd64

Server:
 Engine:
  Version:      17.12.1-ce
  API version:  1.35 (minimum version 1.12)
  Go version:   go1.9.4
  Git commit:   cdade7ddb4
  Built:        Fri Mar 29 11:16:05 2019
  OS/Arch:      linux/amd64
  Experimental: true
```

**Output of `docker info`:**

on 516 
```
[root@tos-516 ~]# docker info
Containers: 76
 Running: 66
 Paused: 0
 Stopped: 10
Images: 19
Server Version: 17.12.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: nvidia runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 8629fabc17c2befd611f03cc257a8b195de8fd7e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 20
Total Memory: 14.18GiB
Name: tos-516
ID: XCGD:XLMH:YGF2:DZQQ:BM6V:CRG2:VICA:HZXW:PWAL:B6R6:RVXQ:7RZY
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 228
 Goroutines: 200
 System Time: 2019-05-21T20:43:26.572798289+08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Registry Mirrors:
 https://172.26.5.16:5000/
Live Restore Enabled: true
```

on 519 
```
[root@tos-519 ~]# docker info
Containers: 42
 Running: 39
 Paused: 0
 Stopped: 3
Images: 20
Server Version: 17.12.1-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: nvidia runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 8629fabc17c2befd611f03cc257a8b195de8fd7e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 20
Total Memory: 14.18GiB
Name: tos-519
ID: WHEP:4UUQ:QZKM:LFJ3:LUJC:MWVH:OZUN:WDCJ:N5D5:Q2H6:5SAU:TTOP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 228
 Goroutines: 196
 System Time: 2019-05-21T20:42:57.662963998+08:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Registry Mirrors:
 https://172.26.5.16:5000/
Live Restore Enabled: true
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

on 516
```
[root@tos-516 ~]# uname -a
Linux tos-516 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux                                                                                                     
[root@tos-516 ~]# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

on 519
```
[root@tos-519 ~]# uname -a
Linux tos-519 3.10.0-327.el7.x86_64 #1 SMP Thu Nov 19 22:10:57 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
[root@tos-519 ~]# cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

**Comment by yuchengwu** [comment]
To fix this on 516, we have to 

1. docker rm all containers running with the broken image
2. docker rmi the broken image
3. docker pull the latest image 

then, docker run works fine.

**Comment by yuchengwu** [comment]
possibly related:
~~https://github.com/moby/moby/issues/22323~~
https://success.docker.com/article/ucp-health-checks-fail-unable-to-find-user-nobody-no-matching-entries-in-passwd-file-observed

**Comment by yuchengwu** [comment]
`standard_init_linux.go:190: exec user process caused "exec format error"` looks like related to an os Arch issue https://github.com/containers/buildah/issues/475

**Comment by thaJeztah** [comment]
Note that this example looks incorrect (`--user` is provided _after_ the image name, so docker will try to run it as the container's process);

```
[root@tos-516 ~]# docker run --rm -ti transwarp/node-exporter:live  --user root
docker: Error response from daemon: linux spec user: unable to find user nobody: no matching entries in passwd file.    
```

Note that generally, I would not recommend bind-mounting `/etc/passwd` into the container; if you want to run the container as the same users as the user running on the host, using the `id -u` approach should work fine (as in; `/etc/passwd` would only be used to lookup the `uid` for a user with the given name). (Also see the discussion on https://github.com/moby/moby/issues/22323, which I see you commented on).

Looking at the `unable to find user nobody` error; does that entry actually existing the container's `/etc/passwd`? (i.e., is that user listed when running `docker run --rm transwarp/node-exporter:live cat /etc/passwd` ?)


Note that your version of CentOS looks to be quite outdated (`3.10.0-327.el7.x86_64` looks to be CentOS 7.2 - CentOS uses a rolling release model, so CentOS 7.2 became EOL when CentOS 7.3 was release in November 2016). In addition, Docker 17.12 also reached EOL a long time ago and has known vulnerabilities (including a [critical CVE](https://nvd.nist.gov/vuln/detail/CVE-2019-5736) in the runc runtime used by that version), so I'd recommend updating to currently maintained versions of both.


Unfortunately, I don't see a way to reproduce the issue you're reporting; the image you're using looks to be a private image.

Are you able to provide a reproducible case (using a standard image, and reproducible on a current version of Docker)? Otherwise I don't think there's anything actionable in this issue.

**Comment by yuchengwu** [comment]
hi, @thaJeztah thanks for your explanation.

> I would not recommend bind-mounting /etc/passwd into the container

The advice is good, and we actually didn't do so in the production env. The example I posted it with binding  /etc/passwd is just a testing(after we found our image cannot run then we google around and followed a post's steps to debug). 

---

> is that user listed when running docker run --rm transwarp/node-exporter:live cat /etc/passwd ?

very good question, we had executed a similar command but got  
```
[root@tos-516 ~]# docker run -it --rm --user nobody transwarp/node-exporter:live cat /etc/passwd
node_exporter: error: unexpected cat, try --help         
```

---

We have been always recommending our clients to upgrading kernel, OS and related softwares to the latest stable version, but due to kinds of reasons(mainly there is a need to compatible with older programs) it might not be accomplished. 

---

Yes, no way to reproduce it is bad and hard to find the root cause, I will try to reproduce it and report it back if I could.

---

Well, I suddenly realize that I might have digressed from the original question, the previous comments I posted are more likely a process of debugging the broken image why cannot be run. 

The original problem is two image with same image id and size run on the same environment but got different behaviors, that is weird, don't you think so?

**Comment by thaJeztah** [comment]
> The original problem is two image with same image id and size run on the same environment but got different behaviors, that is weird, don't you think so?

Yes, that should not be the case; possibly one machine is in a certain state that's causing this (could also be an issue with overlayfs; overlayfs on RHEL and CentOS versions before 7.4 still had certain issues (RHEL 7.4 officially started supporting it for container filesystems; https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html-single/7.4_release_notes/index#new_features_file_systems)

SELinux could cause issues blocking access to `/etc/passwd` (which could lead to the container not being able to read the file)

**Comment by yuchengwu** [comment]
Ah, thanks for the information, I just checked the hosts with `getenforce` and SElinux is in `Disabled` state. Not sure if anybody has changed it , but if this is the case then it make sense to me.

However, I will follow this issue up and see if I can find more info.

**Comment by olljanat** [comment]
@yuchengwu this sounds very similar issue than what I reported on https://github.com/moby/moby/issues/38488#issuecomment-492801678 and which will be fixed after #39193 gets merged.

To be sure about that you need check `/etc/passwd` file content from those containers.

**Comment by yuchengwu** [comment]
hi @olljanat , thanks the info. https://github.com/moby/moby/issues/38488 is what I had experienced, you remind me that I had done similar steps,
```
[root@tos-516 ~]# docker rmi -f transwarp/node-exporter:live                                                                                                                                               
Untagged: transwarp/node-exporter:live                                                                                                                                                                     
Untagged: transwarp/node-exporter@sha256:323ccd3f3f3ee35aa566cb02cc359326dd1851cac6a11749e2456d60f87fa08a                                                                                                  
Deleted: sha256:bbae768f602b6e4d2433b781dc889f7d4811f03bb8599ee01bb69b7b2223a299
[root@tos-516 ~]# docker images | grep transwarp/node-exporter                                                                                                                                             
[root@tos-516 ~]# docker images | grep transwarp/node-exporter
[root@tos-516 ~]# docker pull transwarp/node-exporter:live
live: Pulling from transwarp/node-exporter
0de338cf4258: Already exists
f508012419d8: Already exists
d764f7880123: Already exists
Digest: sha256:323ccd3f3f3ee35aa566cb02cc359326dd1851cac6a11749e2456d60f87fa08a
Status: Downloaded newer image for transwarp/node-exporter:live
```
 docker rmi image is OK but the image layers still on disk when I re-pull. It looks like image cannot be real deleted from disk unless there is no other references(e.g., referenced by a running or dead container)


Yes, I will check the next time.

**Comment by olljanat** [comment]
Yea issue on current implementation is that Docker daemon assumes that it have all the power to be able to fully remove layers which is wants and that is not always true.

That why I want to now change that behavior like this: https://github.com/moby/moby/pull/39193#issuecomment-494982865
But that need still some discussion as I'm touching quite old / core logic of layer handling (it was introduced on #17924) 

I noticed that you are using XFS file system (which is default on CentOS). I'm not familiar with that one but I assume that it have similar capabilities than Windows file system that it actually can lock files on way that Docker always cannot remove files and reason why this have not been seen by more users is that those EXT3/4 file systems are more common (my assumption).

**Comment by zq-david-wang** [comment]
> [root@tos-516 ~]# docker run -it --rm --user nobody transwarp/node-exporter:live cat /etc/passwd
node_exporter: error: unexpected cat, try --help   

@yuchengwu  you're running it wrong,  you need to change entrypoint, otherwise you would run your entrypoint in your image, try this:
docker run -it --rm --user nobody --entrypoint=/bin/bash  transwarp/node-exporter:live -- -c  "cat /etc/passwd"

And there is possibility you host has read/write inconsistancy issue for memory/disk, better to make memtest as well

**Comment by yuchengwu** [comment]
sorry for the late response, occupied by many other things, also it took me a while to understand what image ID is.

@olljanat , the issue was most likely caused by some broken image layers.
I love the idea https://github.com/moby/moby/pull/39193#issuecomment-494982865, hoping it can solve my case. 
Meanwhile one thing I want to confirm, layers marked as invalid should have a modified content? Otherwise if user re-downloaded the layer from remote the two layers'(local and remote) sha256sum would collide.

I am not familiar with XFS either. I asked our colleague why XFS instead of EXT3/4, the answer is we encountered some problems(instability, start docker get warning) when deploy docker on EXT3/4.

----

@zq-david-wang thank you very much for helping debugging and pointing me in the right direction!! I tried the command but got 

```
[root@tos-516 _output]# docker run -it --rm --user nobody --entrypoint=/bin/bash transwarp/node-exporter:live -- -c "cat /etc/passwd"
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/bin/bash\": stat /bin/bash: no such file or directory": unknown.
```

while using /bin/sh entrypoint works fine

```
[root@tos-516 _output]# docker run -it --rm --user nobody --entrypoint=/bin/sh  transwarp/node-exporter:live 
/ $ cat /etc/passwd
root:x:0:0:root:/root:/bin/sh
daemon:x:1:1:daemon:/usr/sbin:/bin/false
bin:x:2:2:bin:/bin:/bin/false
sys:x:3:3:sys:/dev:/bin/false
sync:x:4:100:sync:/bin:/bin/sync
mail:x:8:8:mail:/var/spool/mail:/bin/false
www-data:x:33:33:www-data:/var/www:/bin/false
operator:x:37:37:Operator:/var:/bin/false
nobody:x:65534:65534:nobody:/home:/bin/false
```

Would you mind expanding a bit on that, like any recommended tools I can use to confirm or rule out this possibility.

**Comment by olljanat** [comment]
> @olljanat , the issue was most likely caused by some broken image layers.
> I love the idea [#39193 (comment)](https://github.com/moby/moby/pull/39193#issuecomment-494982865), hoping it can solve my case.
> Meanwhile one thing I want to confirm, layers marked as invalid should have a modified content? Otherwise if user re-downloaded the layer from remote the two layers'(local and remote) sha256sum would collide.

Those does not collide because layers are stored to path `/var/lib/docker/overlay2/<random id>` and that random part will change when layer is re-downloaded.

However I'm not still 100% sure if you are facing same issue. We need to get that `cat /etc/passwd` output from both working and broken node to confirm that.

> I am not familiar with XFS either. I asked our colleague why XFS instead of EXT3/4, the answer is we encountered some problems(instability, start docker get warning) when deploy docker on EXT3/4.

OK. It would be nice if you can verify what there is open issue here about that one and or if those are fixed already on recent versions?

**Comment by zq-david-wang** [comment]
A simple, but no thorough, way to make memory test is by checking checksum of a big file several times and verify whether the yielding is consistent
1. dd if=/dev/urandom of=bfile bs=4194304 count=[n]
This will create a random file of 4*nMB, you need to make sure that 4*n is large than your host's total memory in MB
2. run `md5sum bfile` several times

**Comment by yuchengwu** [comment]
I come back with a new finding: docker run will not check the integrity of the layer data, though content addressable hash can protect image from being tempered during pulling and pushing.

Let's see an example, before starting be familiar with layer, layer diffID, layer digest, layer chainID, image ID(https://github.com/opencontainers/image-spec/blob/master/config.md), image digest would be helpful.
Lines start with `#` are added comments

step 1: pull image hello-world

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker pull hello-world
Using default tag: latest
latest: Pulling from library/hello-world
Digest: sha256:0e11c388b664df8a27a901dce21eb89f11d8292f7fca1b3e3c4321bf7897bffe
Status: Image is up to date for hello-world:latest
```

step 2: check image ID and image digest

OK, we get two different DIGESTs, this should not happen I think, let's put it aside now as it is another story not scoped in this issue.
```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker images --digests hello-world
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
hello-world         latest              sha256:0e11c388b664df8a27a901dce21eb89f11d8292f7fca1b3e3c4321bf7897bffe   fce289e99eb9        4 months ago        1.84kB
hello-world         latest              sha256:92c7f9c92844bbbb5d0a101b22f7c2a7949e40f8ea90c8b3bc396879d95e899a   fce289e99eb9        4 months ago        1.84kB
```

step 3: verify docker runs happily

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker run hello-world 

Hello from Docker!
This message shows that your installation appears to be working correctly.

To generate this message, Docker took the following steps:
 1. The Docker client contacted the Docker daemon.
 2. The Docker daemon pulled the "hello-world" image from the Docker Hub.
    (amd64)
 3. The Docker daemon created a new container from that image which runs the
    executable that produces the output you are currently reading.
 4. The Docker daemon streamed that output to the Docker client, which sent it
    to your terminal.

To try something more ambitious, you can run an Ubuntu container with:
 $ docker run -it ubuntu bash

Share images, automate workflows, and more with a free Docker ID:
 https://hub.docker.com/

For more examples and ideas, visit:
 https://docs.docker.com/get-started/
```

step 4: find the image content file

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo find /var/lib/docker -name fce289e99eb9*
/var/lib/docker/image/aufs/imagedb/content/sha256/fce289e99eb9bca977dae136fbe2a82b6b7d4c372474c9235adc1741675f587e
```

step 5: check its layer diff id

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo cat /var/lib/docker/image/aufs/imagedb/content/sha256/fce289e99eb9bca977dae136fbe2a82b6b7d4c372474c9235adc1741675f587e | jq .rootfs
{
  "type": "layers",
  "diff_ids": [
    "sha256:af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3"
  ]
}
```

step 6: calculate chain id from diff id, only one bottom most layer here, so chain id0 = diff id0, then find chain id0 location

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo find /var/lib/docker -name af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3* 
# file contains  layer digests
/var/lib/docker/image/aufs/distribution/v2metadata-by-diffid/sha256/af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3
# each layer has such a directory to store layer info, next step we will dive into this dir
/var/lib/docker/image/aufs/layerdb/sha256/af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3
```

step 7: dive into chain id directory

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo ls /var/lib/docker/image/aufs/layerdb/sha256/af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3 -l                                     
total 16
# a random UUID generated when downloading the layer and points to the location where the current layer data stored
-rw-r--r-- 1 root root  64 May 29 18:08 cache-id
# stores diff-id
-rw-r--r-- 1 root root  71 May 29 18:08 diff
# layer size in bytes
-rw-r--r-- 1 root root   4 May 29 18:08 size
# don't know what is used for yet.
-rw-r--r-- 1 root root 223 May 29 18:08 tar-split.json.gz
```

step 8: find the cache-id location

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo cat /var/lib/docker/image/aufs/layerdb/sha256/af0b15c8625bb1938f1d7b17081031f649fd14e6b233688eea3c5483994a66a3/cache-id
747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893

wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo find /var/lib/docker -name 747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893* 
/var/lib/docker/aufs/layers/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893 # records the relationship of the layer and its ancestors
/var/lib/docker/aufs/mnt/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893 # the merged layer data with all its ancestors' layers
/var/lib/docker/aufs/diff/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893 # the current layer data(a change set)
```

step 9: find the hello binary then do some modification

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo ls /var/lib/docker/aufs/diff/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893
hello

wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo mv /var/lib/docker/aufs/diff/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893/hello{,.bak}
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ sudo ls /var/lib/docker/aufs/diff/747051d08398cd12914637e2da500f6dd41d9bfa582605be187047a54f39a893
hello.bak
```

now, the cache-id points to a broken layer, what will happened if we execute docker run hello-world again

step 10: docker run hello-world

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker run hello-world
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/hello\": stat /hello: no such file or directory": unknown.
ERRO[0001] error waiting for container: context canceled 
```

Ops, seems that docker will not check whether the layer content was tampered or not

step 11: delete the local image

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker rmi hello-world && docker pull hello-world
Error response from daemon: conflict: unable to remove repository reference "hello-world" (must force) - container 9fae78c4ae7c is using its referenced image fce289e99eb9
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker rmi -f hello-world 
Untagged: hello-world:latest
Untagged: hello-world@sha256:0e11c388b664df8a27a901dce21eb89f11d8292f7fca1b3e3c4321bf7897bffe
Untagged: hello-world@sha256:92c7f9c92844bbbb5d0a101b22f7c2a7949e40f8ea90c8b3bc396879d95e899a
Deleted: sha256:fce289e99eb9bca977dae136fbe2a82b6b7d4c372474c9235adc1741675f587e
```

step 12: pull the latest image and run again

```
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker pull hello-world 
Using default tag: latest
latest: Pulling from library/hello-world
Digest: sha256:0e11c388b664df8a27a901dce21eb89f11d8292f7fca1b3e3c4321bf7897bffe
Status: Image is up to date for hello-world:latest
wyc@smile:~/etcdlog/warp/33917/helle-world/tmp2$ docker run hello-world
docker: Error response from daemon: OCI runtime create failed: container_linux.go:348: starting container process caused "exec: \"/hello\": stat /hello: no such file or directory": unknown.
ERRO[0001] error waiting for container: context canceled
```

still doesn't work.

Summary: apparently docker was not aware of locate modification of layer content, I personally believe this is not intended.

**Comment by olljanat** [comment]
@yuchengwu I noticed when I was debugging #38488 and docker only calculates digest after download before it unzip layers (I explained that already two weeks ago on https://github.com/moby/moby/issues/38488#issuecomment-492801678 :wink: ) That looks to be by design.

Only way to prevent that is setup content trust: https://docs.docker.com/engine/security/trust/content_trust/

However, I noticed that one very important detail about your setup is missing on here. Are your docker nodes physical or virtual servers? And if they are virtual are they running on different physical nodes?

As @zq-david-wang was assuming that they might be memory or disk issue but those are really valid cases only if you are running on non-virtualized environment or if all your virtual machines are running top of same physical server.

**Comment by olljanat** [comment]
@yuchengwu and just to clarify. For me your issue looks same as #38488 and if that is case then #39193 will fix the issue.

**BUT** we need to remember that this part of code have been there almost five years and there is lot of users who have not seen this issue so we need to make sure that I'm not over assuming and issue is not somewhere else. 

We have seen that because we run Windows containers on production and it looks to be that there is no some many others who are crazy enough to do so...

**Comment by yuchengwu** [comment]
thanks for your quick response
> @yuchengwu I noticed it when I was debugging #38488 and docker only calculates digest after download before it unzip layers (I explained that already two weeks ago on #38488 (comment) wink ) That looks to be by design.

I also learned this is by design, my point is why there is not a mechanism to verify unziped layer integrity, as a container rely on those layers to constructed its rootfs which should be always immutable.

I will take time to check out  https://docs.docker.com/engine/security/trust/content_trust/.

The original issue happened on virtual servers, not sure if they(node 516, 519) are running on same or different physical nodes, needs confirm from my workmate.

---

> BUT we need to remember that this part of code have been there almost five years ...

Aha, so you are touching something really old, be patient and good luck! :wink: 

> We have seen that because we run Windows containers on production ...

Crazy and challenging, and enjoy :beer:

**Comment by olljanat** [comment]
Btw. This one was most likely fixed by #39193 which was released as part of Docker 20.10.0

**Comment by thaJeztah** [comment]
Let me go ahead and close this one
